BUG/ENH: stats: Cauchy distribution fixes

WarrenWeckesser opened the pull request:
* Eliminate spurious overflow warnings from `cauchy.pdf(x)` when |x| > 1.34e154 (i.e. `sqrt(np.finfo(np.float64).max)`).
* Implement `cauchy._logpdf` to extend the input range over which `cauchy.logpdf` gives correct results.  Before this change, `cauchy.logpdf(x)` would return -inf for |x| > 1.34e154.

WarrenWeckesser committed:
BUG/ENH: stats: Cauchy distribution fixes:

* Eliminate spurious overflow warnings from cauchy.pdf(x) when
  |x| > 1.34e154 (i.e. sqrt(np.finfo(np.float64).max)).
* Implement cauchy._logpdf to extend the input range over which
  cauchy.logpdf gives correct results.  Before this change,
  cauchy.logpdf(x) would return -inf for |x| > 1.34e154.

WarrenWeckesser committed:
TST: stats: Tests for cauchy.pdf and cauchy.logpdf.

fancidev commented:
Good stuff!

One question: assuming `x > 0`, does the expression `np.log(x) + np.log(1/x + x)` introduce (non-negligible) cancellation error for small and large `x`? For example, when `x` is large, `1/x` becomes small and `log(1/x+x)` is close to `log(1/x)  == -log(1/x)`.

fancidev commented on scipy/stats/_continuous_distns.py:
```diff
@@ -1451,7 +1451,22 @@ def _shape_info(self):
+        # The formula
+        #     log(1/(pi*(1 + x**2))) = -log(pi) - log(1 + x**2)
+        #                            = -log(pi) - log(|x|*(1/|x| + |x|))
+        #                            = -log(pi) - (log(|x| + log(1/|x| + |x|)))
```
Typo: one closing parentheses is misplaced

WarrenWeckesser replied:
Thanks. The comment has been rewritten in the latest update.

fancidev commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -509,6 +509,28 @@ def test_cdf_ppf(self):
+         (2e18, -85.42408759475494),
+         (1e200, -922.1787670834676),
+         (4e303, -1399.283884962481)])
+    def test_logpdf(self, x, ref):
```
Is it worth including `x=np.finfo(float).smallest_subnormal`, `smallest_normal`, and `largest` in the test cases, for completeness?

WarrenWeckesser commented:
> One question: assuming x > 0, does the expression np.log(x) + np.log(1/x + x) introduce (non-negligible) cancellation error for small x?

Yes it does! Cancellation occurs when |x| < 1, and the amount of cancellation increases as |x| → 0.  I was concerned about that, but my initial tests showed that the precision was pretty good, despite the reliance on cancellation to compute the result.  One reason, as you noted, is that for small enough x, the absolute error of the difference of the logs is small enough that the overall relative error of the sum when `_LOG_PI` is included is small.

Still, your question prompted me to experiment a bit more, and I found that, although the expression that I used wasn't bad, better precision can be obtained for |x| < 1 by using `-_LOG_PI - np.log1p(absx**2)`.  This can't be used for large |x| because of the possibility of overflow.  So I rewrote the method, begrudgingly using `_lazywhere`; in the tradeoff between precision and speed, I'm leaning towards precision here.  For |x| > 1, I changed the expression to `-_LOG_PI - (2*np.log(absx) + np.log1p(1/absx**2))`, because here, too, I found this version has slightly (but only very slightly) better precision than what I originally used.

fancidev commented on scipy/stats/_continuous_distns.py:
```diff
@@ -1451,7 +1451,26 @@ def _shape_info(self):
+            y = _lazywhere(absx < 1, (absx,),
+                           f=lambda absx: -_LOG_PI - np.log1p(absx**2),
+                           f2=lambda absx: (-_LOG_PI -
+                                            (2*np.log(absx) + np.log1p(1/absx**2))))
```
The `with errstate` guard could be eliminated if `1/x**2` is replaced by (`1/x)**2`, but the current code also works without problem!

WarrenWeckesser replied:
Yup, good idea.  I included this in the recent update.

fancidev commented:
Looks perfect to me!

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -1451,7 +1451,24 @@ def _shape_info(self):
+        #                            = -log(pi) - log(x**2*(1 + 1/x**2))
+        #                            = -log(pi) - (2log(|x|) + log1p(1/x**2))
+        # are used here.
+        absx = np.abs(x)
```
I checked, and it's probably best to keep this as you have it - always calculate `absx` so it can be used in the `cond` and in `f2`. The alternative - `cond = (x < 1) & (x > -1)` and calculating `np.abs(x)` only where needed - is not clearly better.

mdhaber reviewed (approved):
This looks fine. Thought I'd leave it open in case you want to adjust the threshold of `cond` in `_lazywhere`, but if you're happy with it as-is, I'll merge.
